app/services/concerns/xccdf/rules.rb

```diff
@@ -9,51 +9,56 @@ module Rules
       def rules_already_saved
         return @rules_already_saved if @rules_already_saved.present?
 
-        rule_ref_ids = @oscap_parser.rule_objects.map(&:id)
+        rule_ref_ids = @op_rules.map(&:id)
         @rules_already_saved = Rule.select(:id, :ref_id)
-                                   .where(ref_id: rule_ref_ids)
+                                   .where(ref_id: rule_ref_ids,
+                                          benchmark_id: @benchmark.id)
                                    .includes(:profiles)
       end
 
-      def add_profiles_to_old_rules(rules, new_profiles)
-        preexisting_profiles = ProfileRule.select(:profile_id)
-                                          .where(rule_id: rules.pluck(:id))
-                                          .pluck(:profile_id).uniq
-        rules.find_each do |rule|
+      def add_profiles_to_old_rules
+        rules_already_saved.find_each do |rule|
           new_profiles.each do |profile|
-            unless preexisting_profiles.include?(profile.id)
+            unless preexisting_profile_ids.include?(profile.id)
               ProfileRule.create(rule_id: rule.id, profile_id: profile.id)
             end
           end
         end
       end
 
+      def preexisting_profile_ids
+        @preexisting_profile_ids ||= ProfileRule.select(:profile_id).where(
+          rule_id: rules_already_saved.pluck(:id)
+        ).pluck(:profile_id).uniq
+      end
+
       def new_rules
         return @new_rules if @new_rules
 
         ref_ids = rules_already_saved.pluck(:ref_id)
-        @new_rules = @oscap_parser.rule_objects.reject do |rule|
+        @new_rules = @op_rules.reject do |rule|
           ref_ids.include? rule.id
         end
       end
 
       def save_rules
-        add_profiles_to_old_rules(rules_already_saved, new_profiles)
+        add_profiles_to_old_rules
         associate_rule_references
         Rule.import!(new_rule_records, recursive: true)
```

Comment:
With an empty database, I'm seeing `ActiveRecord::RecordInvalid: Validation failed: Rule references is invalid` on this database. After that error, if I rerun the task, the error goes away (I think this happens because rules were already imported properly, so `new_rule_records` is `[]` the 2nd time we run the import, but rule references were lost)